apollo-federation/src/sources/connect/json_selection/selection_trie.rs

```diff
@@ -0,0 +1,390 @@
+}
+
+impl Display for SelectionTrie {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut need_space = false;
+
+        for (key, sub) in self.selections.iter() {
+            if need_space {
+                write!(f, " ")?;
+            }
+
+            if sub.is_empty() {
+                if sub.is_used() {
+                    write!(f, "{}", quote_if_necessary(key))?;
+                    need_space = true;
+                }
+            } else {
+                write!(f, "{} {{ {} }}", quote_if_necessary(key), sub)?;
+                need_space = true;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl PartialEq for SelectionTrie {
+    fn eq(&self, other: &Self) -> bool {
+        self.used == other.used && self.selections == other.selections
+    }
+}
+
+impl Hash for SelectionTrie {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.selections
+            .iter()
+            .fold(0, |acc, (key, sub)| {
+                let mut hasher = std::hash::DefaultHasher::default();
+                (key, sub).hash(&mut hasher);
+                acc ^ hasher.finish()
+            })
+            .hash(state);
+    }
+}
+
+impl SelectionTrie {
+    pub(crate) fn new() -> Self {
+        Self {
+            used: false,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn new_used() -> Self {
+        Self {
+            used: true,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.selections.is_empty()
+    }
+
+    pub(crate) fn keys(&self) -> impl Iterator<Item = &String> {
+        self.selections.keys()
+    }
+
+    pub(crate) fn get(&self, key: impl Into<String>) -> Option<&SelectionTrie> {
+        self.selections.get(&key.into()).map(|sub| sub.as_ref())
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = (&str, &SelectionTrie)> {
+        self.selections
+            .iter()
+            .map(|(key, sub)| (key.as_str(), sub.as_ref()))
+    }
+
+    pub(crate) fn key_ranges(&self, key: &str) -> impl Iterator<Item = Range<usize>> {
+        self.key_ranges
+            .get(key)
+            .into_iter()
+            .flat_map(|ranges| ranges.iter())
+            .cloned()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn has_str_path<'a>(&self, path: impl IntoIterator<Item = &'a str>) -> bool {
+        let mut current = self;
+        for key in path {
+            if let Some(sub) = current.get(key) {
+                current = sub;
+            } else {
+                return false;
+            }
+        }
+        current.is_used()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn add_str_path<'a>(
+        &mut self,
+        path: impl IntoIterator<Item = &'a str>,
+    ) -> &mut Self {
+        path.into_iter()
```

Comment:
maybe replace with https://doc.rust-lang.org/std/iter/trait.Extend.html

Reply:
I'm already using `SelectionTrie::extend(&mut self, other: &SelectionTrie) -> &mut Self` for (destructively) merging `other` into `self`, and think I'm happy with the clarity of `add_str_path`, as it stands. It's an easy method to test, but it's not really the workhorse of this system, because we generally want to have a chance of collecting location information along with the field names stored in the trie, and you can only provide the field names themselves with `add_str_path`.